CIP-????/README.md

```diff
@@ -0,0 +1,294 @@
+functions, these are:
+
+1. A verification key (length depends on the algorithm chosen);
+2. An input to verify (either the message itself, and a hash);
+3. A signature.
+
+The two different schemes handle deserialization internally: specifically, there
+is a distinction made between 'external' representations, which are expected as
+arguments, and 'internal' representations, used only by the implementations
+themselves. This creates different expecations for each argument for both of
+these schemes; we describe these below.
+
+For the [all signature
+schemes](),
+the keys are required to be given in the same format as NIST test vectors.
+
+The builtin operations will error with a descriptive message if given inputs
+that don't correspond to the constraints above, return `False` if the signature
+fails to verify the input given the key, and `True` otherwise.
+
+### Post-quantum wallets
+
+We should permit using all chosen signature algorithms to lock transactions.
+
+This allows an _opt-in_ migration as more and more wallet apps will support new signature schemes.
+
+For extra safety, users may choose to use conjunction of wallet signatures as they
+are already implemented in native ledger.
+
+Note that default would be to query wallet addresses generated by hierarchical wallet schema
+for each scheme.
+For the signing transactions and data, function should default for existing signature
+algorithm to preserve backwards compatibility.
+
+There should be additional attributes as described by [CIP-0030](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0030):
+
+```
+cardano.{walletName}.signAlgo
+cardano.{walletName}.hashAlgo
+```
+
+There should be additional extension registered if [CIP on wallet extensibility](https://github.com/cardano-foundation/CIPs/pull/462).
+
+Wallet providers may choose to automatically generate ultra-secure wallets for multiple quantum signature
+algorithms from a single seed. In such case, they are encouraged to provide a visible
+hint of what is the current preferred.
+
+### Use of standard hash algorithm
+
+Note that use of hash would require use of quantum-resistant hash algorithm.
+[SHA2, SHA3, BLAKE2 are considered quantum-safe](https://cryptobook.nakov.com/quantum-safe-cryptography#quantum-safe-and-quantum-broken-crypto-algorithms),
+in particular standard 256-bit hash used in Cardano [should be safe for a long time](https://cryptobook.nakov.com/quantum-safe-cryptography#quantum-safe-and-quantum-broken-crypto-algorithms).
+This should be documented in the Plutus documentation of `verifySignature`.
+
+However, in the interest of long-term API evolution we propose to add a flexible hashing interface:
+
+```
+data SignAlgo = SHA2_256
+              | SHA3_256
+              | Blake2
+
+hashWithAlgo :: HashAlgo -> BuiltinByteString -> BuiltinByteString
+```
+
+This will also allow for gradual _opt-in_ migration of scripts as the blockchain evolves.
```

Comment:
Again, this is about extensibilty, and I still think it should be clearly separated out and argued for.